### ISSUE TITLE
"input list" selection keys

### ISSUE THREAD

**Issue by ineu** [nushell/nushell#9911]
### Related problem

"input list" can be used as a replacement for fzf/skim. Alas, the only way to select an item from the list is using the cursor keys on the keyboard.

### Describe the solution you'd like

Make "input list" work with ctrl+n/ctrl+p for the next/previous option in the list. This is how nushell itself works, and is very common in general, used in emacs and all readline-based CLIs.

**Comment by amtoine** [comment]
@ineu 
internally, `input list` uses the [dialoguer](https://crates.io/crates/dialoguer) crate.
the thing is i do not see any mention to key bindings in either the source of `input list`, nor the documentation of `dialoguer` :thinking: 

if that's a limitation from the library, it might not be possible to achieve this right now :confused:

**Comment by ineu** [comment]
Ah, that's unfortunate. Thanks for clarification.
Just out of curiosity: could it be feasible to make `input list` use `reedline` instead?

**Comment by amtoine** [comment]
> Ah, that's unfortunate.
>
yeah `input list` is far from perfect but it's such a good start to have a pure Nushell `fzf` replacement :star_struck: 

> Thanks for clarification.
>
anytime :blush: 

> Just out of curiosity: could it be feasible to make `input list` use `reedline` instead?
>
what do you by "*using `reedline` instead*"?

**Comment by ineu** [comment]
Just thinking aloud, maybe I'm missing something.

So nushell already uses the `reedline` library. It's not a menu builder, but it can handle `ctrl+?` keys, and return the input. So `input list` could use it to get the input, and then the duty of `input list` would be to
1. Rerender the output (if a user pressed ctrl+n, the cursor position must be updated in the selection menu)
2. Return the final input to the user

So I wonder if there's anything else `dialoguer` does which cannot be done with described approach?

**Comment by amtoine** [comment]
yeah i see what you mean :ok_hand: 

i'm not sure at all :thinking: 
i'd definitely review PRs trying to make `input list` better :innocent:

**Comment by fdncred** [comment]
It may not be ideal, but you can also type to find a selection. e.g. `ls | input list -f "pick something"` allows you to type the name of the file and hit enter for the selection. I agree that you cannot move the `>` indicator without the arrow keys.

**Comment by amtoine** [comment]
> The main problem with the cursor keys is that they enforce you to take your hands off the home row, which is exactly what vim/emacs keybindings try to avoid.
>
i cheat with my keyboard and can transform my hjkl keys into the arrow keys with a keys under my thumb :smirk: :wink: 

but i agree about leaving the home row in a keyboard-focus workflow is annoying :+1:

**Comment by NotLebedev** [comment]
It's not impossible to do this using `input listen` in pure nushell. Something like this:
```nu
export def "input select" [] {
  let data = $in
  let data_length = $data | length
  mut current = 0
  mut first_iter = true
  loop {
    mut output = "" # Buffering output to limit flickering

    if not $first_iter {
      $output = ([$output $"\e[($data_length + 1)A"] | str join)
    }
    
    for e in $data --numbered {
      if $e.index == $current {
        $output = ([$output $"> ($e.item)\n"] | str join)
      } else {
        $output = ([$output $"  ($e.item)\n"] | str join)
      }
    }

    if $first_iter {
      $first_iter = false
    }
    print $output
    
    let key = input listen -t [key]
    match $key {
      {
        type: key 
        key_type: char
        code: p
        modifiers: [ control ]
      } => {
        $current = ([($current - 1) 0] | math max)
      }
      {
        type: key 
        key_type: char
        code: n
        modifiers: [ control ]
      } => {
        $current = ([($current + 1) ($data_length - 1)] | math min)
      }
      {
        type: key
        key_type: other
        code: enter
        modifiers: []
      } => { return ($data | get $current) }
    }
  }
}
```

**Comment by fdncred** [comment]
very cool @NotLebedev!
